File: spec/cpp_stl/test_floating_points.cpp
#include <boost/test/unit_test.hpp>

#include <floating_points.h>

#include <iostream>
#include <fstream>
#include <vector>

BOOST_AUTO_TEST_CASE(test_floating_points) {
    std::ifstream ifs("src/floating_points.bin", std::ifstream::binary);
    kaitai::kstream ks(&ifs);
    floating_points_t* r = new floating_points_t(&ks);

    double delta = 1e-5;

    BOOST_CHECK_EQUAL(r->single_value(), 0.5f);
    BOOST_CHECK_EQUAL(r->single_value_be(), 0.5f);

    BOOST_CHECK_EQUAL(r->double_value(), 0.25);
    BOOST_CHECK_EQUAL(r->double_value_be(), 0.25);

    BOOST_CHECK_CLOSE(r->approximate_value(), 1.2345, delta);

    BOOST_CHECK_CLOSE(r->single_value_plus_int(), 1.5f, delta);
    BOOST_CHECK_CLOSE(r->single_value_plus_float(), 1.0, delta);
    BOOST_CHECK_CLOSE(r->double_value_plus_float(), 0.3, delta);

    delete r;
}

File: spec/construct/test_floating_points.py
import unittest

from floating_points import _schema

class TestFloatingPoints(unittest.TestCase):
    def test_floating_points(self):
        r = _schema.parse_file("src/floating_points.bin")

        prec = 6

        self.assertEqual(r.single_value, 0.5)
        self.assertEqual(r.single_value_be, 0.5)
        self.assertEqual(r.double_value, 0.25)
        self.assertEqual(r.double_value_be, 0.25)

        self.assertAlmostEqual(r.approximate_value, 1.2345, prec)

        self.assertAlmostEqual(r.single_value_plus_int, 1.5, prec)
        self.assertAlmostEqual(r.single_value_plus_float, 1.0, prec)
        self.assertAlmostEqual(r.double_value_plus_float, 0.3, prec)
